File: LeetCode/2147_NumberOfWaysToDivideALongCorridor/lc2147.cpp
#include <cassert>
#include <string>

class Solution {
public:
  int numberOfWays(std::string corridor) {
    const int MOD = 1'000'000'007;
    long ans = 1;
    int lastSeat = -1;
    int totalSeats = 0;
    for (int i = 0; i < corridor.size(); i++) {
      if ('S' == corridor[i]) {
        ++totalSeats;
        if ((totalSeats & 1) && (totalSeats >= 3)) {
          ans = (ans * (i - lastSeat)) % MOD;
        }
        lastSeat = i;
      }
    }

    // Corner case: if totalSeats is even, there is no way to put the odd one to use.
    // Corner case: if there are no seats at all, there can be no segments at all.
    if ((totalSeats & 1) == 1 || totalSeats == 0) {
      ans = 0;
    }

    return ans;
  }
};

int main() {
  Solution s;
  assert(s.numberOfWays("SSPPSPS") == 3);
  assert(s.numberOfWays("PPSPSP") == 1);
  assert(s.numberOfWays("S") == 0);
}

File: LeetCode/2147/golang/lc2147.go
package lc2147

// Runtime complexity: O(n)
// Auxiliary space: O(1)
// Subjective level: medium+
// Solved on: 2025-12-14
func numberOfWays(corridor string) int {
	const MOD64 int64 = 1_000_000_007
	lastSeatIdx := -1
	totalSeats := 0
	ans := int64(1)

	for i, elem := range corridor {
		if elem == 'S' {
			totalSeats += 1
			if totalSeats%2 == 1 && totalSeats >= 3 {
				ans = (ans * int64(i-lastSeatIdx)) % MOD64
			}
			lastSeatIdx = i
		}
	}

	// Corner case #01: if there are no seats at all, there can be no segments at all.
	// Corner case #02: if totalSeats is even, there is no way to put the odd one to use.
	// both these values are known only at the end of a full scan of the array.
	if totalSeats == 0 || totalSeats%2 == 1 {
		return 0
	}

	return int(ans)
}
